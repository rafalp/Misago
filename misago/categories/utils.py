from ..acl.objectacl import add_acl_to_obj
from ..readtracker import categoriestracker
from .models import Category


def get_categories_tree(
    request, parent=None, join_posters=False
):  # pylint: disable=too-many-branches
    if not request.user_acl["visible_categories"]:
        return []

    if parent:
        queryset = parent.get_descendants().order_by("lft")
    else:
        queryset = Category.objects.all_categories()

    queryset_with_acl = queryset.filter(id__in=request.user_acl["visible_categories"])
    if join_posters:
        queryset_with_acl = queryset_with_acl.select_related("last_poster")

    visible_categories = list(queryset_with_acl)

    categories_dict = {}
    categories_list = []

    parent_level = parent.level + 1 if parent else 1

    for category in visible_categories:
        category.subcategories = []
        categories_dict[category.pk] = category
        categories_list.append(category)

        if category.parent_id and category.level > parent_level:
            categories_dict[category.parent_id].subcategories.append(category)

    add_acl_to_obj(request.user_acl, categories_list)
    categoriestracker.make_read_aware(request, categories_list)

    for category in reversed(visible_categories):
        if category.acl["can_browse"]:
            category.parent = categories_dict.get(category.parent_id)
            if category.parent:
                category.parent.threads += category.threads
                category.parent.posts += category.posts

                if category.parent.last_post_on and category.last_post_on:
                    parent_last_post = category.parent.last_post_on
                    category_last_post = category.last_post_on
                    update_last_thread = parent_last_post < category_last_post
                elif not category.parent.last_post_on and category.last_post_on:
                    update_last_thread = True
                else:
                    update_last_thread = False

                if update_last_thread:
                    category.parent.last_poster = category.last_poster
                    category.parent.last_post_on = category.last_post_on
                    category.parent.last_thread_id = category.last_thread_id
                    category.parent.last_thread_title = category.last_thread_title
                    category.parent.last_thread_slug = category.last_thread_slug
                    category.parent.last_poster_name = category.last_poster_name
                    category.parent.last_poster_slug = category.last_poster_slug

                if not category.is_read:
                    category.parent.is_read = False

    flat_list = []
    for category in categories_list:
        if category.level == parent_level:
            flat_list.append(category)
    return flat_list


def get_category_path(category):
    if category.special_role:
        return [category]

    category_path = []
    while category and category.level > 0:
        category_path.append(category)
        category = category.parent
    return [f for f in reversed(category_path)]
