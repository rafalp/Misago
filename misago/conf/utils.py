from . import hydrators


def get_setting_value(setting):
    return hydrators.hydrate_value(setting.python_type, setting.dry_value)


def set_setting_value(setting, new_value):
    if new_value is not None:
        setting.dry_value = hydrators.dehydrate_value(setting.python_type, new_value)
    else:
        setting.dry_value = None
    return setting.value
