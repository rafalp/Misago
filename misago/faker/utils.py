from functools import wraps
from time import sleep

from django.db import IntegrityError
from django.db.transaction import TransactionManagementError


def retry_on_db_error(f):
    @wraps(f)
    def wrapped_fake_function(*args, **kwargs):
        while True:
            try:
                return f(*args, **kwargs)
            except (IntegrityError, TransactionManagementError):
                sleep(0.2)

    return wrapped_fake_function
