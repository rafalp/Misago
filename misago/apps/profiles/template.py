from datetime import timedelta
from django.conf import settings
from django.template import RequestContext as DjangoRequestContext
from django.utils import timezone
from django.utils.importlib import import_module
from misago.models import User

def RequestContext(request, context=None):
    if not context:
        context = {}
    context['fallback'] = request.path
        
    # Find out if we ignore or follow this user
    context['follows'] = False
    context['ignores'] = False
    if request.user.is_authenticated() and request.user.pk != context['profile'].pk:
        context['follows'] = request.user.is_following(context['profile'])
        context['ignores'] = request.user.is_ignoring(context['profile'])
    
    # Find out if this user allows us to see his activity
    if request.user.pk != context['profile'].pk:
        if context['profile'].hide_activity == 2:
            context['hidden'] = True
        if context['profile'].hide_activity == 1:
            context['hidden'] = context['profile'].is_following(request.user)
    else:
        context['hidden'] = False

    # Find out if this user is online:
    if request.user.pk != context['profile'].pk:
        try:
            context['online'] = context['profile'].sessions.filter(admin=False).filter(last__gt=(timezone.now() - timedelta(minutes=10))).order_by('-last')[0:1][0]
        except IndexError:
            context['online'] = False
    else:
        # Fake "right now" time
        context['online'] = {'last': timezone.now()}

    # Sync member
    if context['profile'].sync_profile():
        context['profile'].save(force_update=True)

    context['tabs'] = []
    for extension in settings.PROFILE_EXTENSIONS:
        profile_module = import_module(extension + '.profile')
        try:
            append_links = profile_module.register_profile_extension(request)
            if append_links:
                for link in append_links:
                    link = list(link)
                    token = link[0][link[0].find('_') + 1:]
                    context['tabs'].append({
                                            'route': link[0],
                                            'active': context['tab'] == token,
                                            'name': link[1],
                                            })
        except AttributeError:
            pass
    return DjangoRequestContext(request, context)
