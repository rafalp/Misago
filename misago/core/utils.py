import hashlib
from datetime import datetime, timedelta

from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.http import Http404
from django.urls import resolve, reverse
from django.utils import html, timezone
from django.utils.encoding import force_str
from django.utils.functional import Promise
from django.utils.module_loading import import_string

MISAGO_SLUGIFY = getattr(settings, "MISAGO_SLUGIFY", "misago.core.slugify.default")

slugify = import_string(MISAGO_SLUGIFY)


def format_plaintext_for_html(string):
    return html.linebreaks(html.urlize(html.escape(string)))


def encode_json_html(string):
    return string.replace("<", r"\u003C")


ISO8601_FORMATS = ("%Y-%m-%dT%H:%M:%S", "%Y-%m-%dT%H:%M:%S.%f")


def parse_iso8601_string(value):
    """turns ISO 8601 string into datetime object"""
    value = force_str(value, strings_only=True).rstrip("Z")

    for format_str in ISO8601_FORMATS:
        try:
            parsed_value = datetime.strptime(value, format_str)
            break
        except ValueError:
            try:
                parsed_value = datetime.strptime(value[:-6], format_str)
                break
            except ValueError:
                pass
    else:
        raise ValueError("failed to hydrate the %s timestamp" % value)

    offset_str = value[-6:]
    if offset_str and offset_str[0] in ("-", "+"):
        tz_offset = timedelta(hours=int(offset_str[1:3]), minutes=int(offset_str[4:6]))
        tz_offset = tz_offset.seconds // 60
        if offset_str[0] == "-":
            tz_offset *= -1
    else:
        tz_offset = 0

    tz_correction = timezone.get_fixed_timezone(tz_offset)
    return timezone.make_aware(parsed_value, tz_correction)


def hide_post_parameters(request):
    """
    Mark request as having sensitive parameters
    We can't use decorator because of DRF uses custom HttpRequest
    that is incompatibile with Django's decorator
    """
    request.sensitive_post_parameters = "__ALL__"


def clean_return_path(request):
    """return path utility that returns return path from referer or POST"""
    if request.method == "POST" and "return_path" in request.POST:
        return _get_return_path_from_post(request)
    return _get_return_path_from_referer(request)


def _get_return_path_from_post(request):
    return_path = request.POST.get("return_path")
    try:
        if not return_path:
            raise ValueError()
        if not return_path.startswith("/"):
            raise ValueError()
        resolve(return_path)
        return return_path
    except (Http404, ValueError):
        return None


def _get_return_path_from_referer(request):
    referer = request.META.get("HTTP_REFERER")
    try:
        if not referer:
            raise ValueError()
        if not referer.startswith(request.scheme):
            raise ValueError()
        referer = referer[len(request.scheme) + 3 :]
        if not referer.startswith(request.META["HTTP_HOST"]):
            raise ValueError()
        referer = referer[len(request.META["HTTP_HOST"].rstrip("/")) :]
        if not referer.startswith("/"):
            raise ValueError()
        resolve(referer)
        return referer
    except (Http404, KeyError, ValueError):
        return None


def is_request_to_misago(request):
    try:
        return request._request_to_misago
    except AttributeError:
        request._request_to_misago = _is_request_path_under_misago(request)
        return request._request_to_misago


def _is_request_path_under_misago(request):
    # We are assuming that forum_index link is root of all Misago links
    forum_index = reverse("misago:index")
    path = request.path

    if len(forum_index) > len(path):
        return False
    return path[: len(forum_index)] == forum_index


def is_referer_local(request):
    referer = request.META.get("HTTP_REFERER")

    if not referer:
        return False
    if not referer.startswith(request.scheme):
        return False
    referer = referer[len(request.scheme) + 3 :]
    if not referer.startswith(request.META["HTTP_HOST"]):
        return False
    referer = referer[len(request.META["HTTP_HOST"].rstrip("/")) :]
    if not referer.startswith("/"):
        return False

    return True


def get_exception_message(exception=None, default_message=None):
    if not exception:
        return default_message

    try:
        exception_message = exception.args[0]
        if isinstance(exception_message, Promise):
            return str(exception_message)
        if isinstance(exception_message, str):
            return exception_message
        return default_message
    except IndexError:
        return default_message


def clean_ids_list(ids_list, error_message):
    try:
        return list(map(int, ids_list))
    except (ValueError, TypeError):
        raise PermissionDenied(error_message)


def get_host_from_address(address):
    if not address:
        return None

    if address.lower().startswith("https://"):
        address = address[8:]
    if address.lower().startswith("http://"):
        address = address[7:]

    address = address.lstrip("/")
    if "/" in address:
        address = address.split("/")[0] or address
    if ":" in address:
        address = address.split(":")[0] or address

    return address


HASH_LENGTH = 8


def get_file_hash(file_obj):
    if not file_obj.size:
        return "0" * HASH_LENGTH
    file_hash = hashlib.md5()
    for chunk in file_obj.chunks():
        file_hash.update(chunk)
    return file_hash.hexdigest()[:HASH_LENGTH]
