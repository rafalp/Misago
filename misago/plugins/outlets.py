from functools import wraps
from typing import Dict, List, Protocol

from django.http import HttpRequest
from django.template import Context
from django.utils.safestring import SafeString, mark_safe

from .enums import PluginOutlet
from .hooks import ActionHook


class PluginOutletHookAction(Protocol):
    def __call__(
        self, request: HttpRequest, context: dict | Context
    ) -> str | SafeString | None:
        pass


class PluginOutletHook(ActionHook[PluginOutletHookAction]):
    __slots__ = ActionHook.__slots__

    def __call__(
        self, request: HttpRequest, context: dict | Context
    ) -> List[str | SafeString | None]:
        return super().__call__(request, context)


template_outlets: Dict[str, PluginOutletHook] = {}


def create_new_outlet(outlet_name: str):
    if outlet_name in template_outlets:
        raise ValueError(f"Template outlet '{outlet_name}' already exists.")

    template_outlets[plugin_outlet.name] = PluginOutletHook()


for plugin_outlet in PluginOutlet:
    create_new_outlet(plugin_outlet.name)


def append_outlet_action(
    outlet_name: str | PluginOutlet, action: PluginOutletHookAction
):
    get_outlet(outlet_name).append_action(action)


def prepend_outlet_action(
    outlet_name: str | PluginOutlet, action: PluginOutletHookAction
):
    get_outlet(outlet_name).prepend_action(action)


def get_outlet(outlet_name: str | PluginOutlet) -> PluginOutletHook:
    try:
        if isinstance(outlet_name, PluginOutlet):
            return template_outlets[outlet_name.name]
        return template_outlets[outlet_name]
    except KeyError as exc:
        raise KeyError(f"Unknown template outlet: {outlet_name}") from exc


def template_outlet_action(f):
    """Decorator for an outlet action that renders a template with returned context."""

    @wraps(f)
    @mark_safe
    def wrapped_outlet_action(request: HttpRequest, context: Context):
        template_data = f(request, context)
        if template_data is None:
            return ""

        if isinstance(template_data, str):
            return _include_template(template_data, context)

        if isinstance(template_data, tuple):
            template_name, new_context = template_data
            return _include_template(template_name, context, new_context)

        return ""

    return wrapped_outlet_action


def _include_template(
    template_name: str, context: Context, new_context: dict | None = None
):
    """Subset of Django include template tag.

    Works only with Django template engine.
    """
    cache = context.render_context.dicts[0].setdefault("_template_outlet_action", {})
    template = cache.get(template_name)
    if template is None:
        template = context.template.engine.get_template(template_name)
        cache[template_name] = template

    if new_context:
        with context.push(**new_context):
            return template.render(context)

    return template.render(context)
