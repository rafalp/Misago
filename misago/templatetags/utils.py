from coffin.template import Library
from haystack.utils import Highlighter
from misago.utils.strings import short_string

register = Library()


@register.object(name='intersect')
def intersect(list_a, list_b):
    for i in list_a:
        if i in list_b:
            return True
    return False


@register.filter(name='short_string')
def make_short(string, length=16):
    return short_string(string, length)


@register.filter(name='highlight')
def highlight_result(text, query, length=500):
    hl = Highlighter(query, html_tag='strong', max_length=length)
    hl = hl.highlight(text)

    return hl