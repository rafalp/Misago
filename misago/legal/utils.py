from hashlib import md5

from django.conf import settings
from django.utils.encoding import force_bytes

from misago.core.cache import cache
from misago.markup import common_flavour

from .models import Agreement, UserAgreement


def set_agreement_as_active(agreement, commit=False):
    agreement.is_active = True
    queryset = Agreement.objects.filter(type=agreement.type).exclude(pk=agreement.pk)
    queryset.update(is_active=False)

    if commit:
        agreement.save(update_fields=['is_active'])
        Agreement.objects.invalidate_cache()


def get_required_user_agreement(user, agreements):
    if user.is_anonymous:
        return None

    for agreement_type, _ in Agreement.TYPE_CHOICES:
        agreement = agreements.get(agreement_type)
        if agreement and agreement['id'] not in user.agreements:
            try:
                return Agreement.objects.get(id=agreement['id'])
            except Agreement.DoesNotExist:
                # possible stale cache
                Agreement.invalidate_cache()
    
    return None


def get_parsed_agreement_text(request, agreement):
    if not agreement.text:
        return None

    cache_name = 'misago_legal_%s_%s' % (agreement.pk, agreement.last_modified_on or '')
    cached_content = cache.get(cache_name)

    unparsed_content = agreement.text

    checksum_source = force_bytes('%s:%s' % (unparsed_content, settings.SECRET_KEY))
    unparsed_checksum = md5(checksum_source).hexdigest()

    if cached_content and cached_content.get('checksum') == unparsed_checksum:
        return cached_content['parsed']
    else:
        parsed = common_flavour(request, None, unparsed_content)['parsed_text']
        cached_content = {
            'checksum': unparsed_checksum,
            'parsed': parsed,
        }
        cache.set(cache_name, cached_content)
        return cached_content['parsed']


def save_user_agreement_acceptance(user, agreement, commit=False):
    user.agreements.append(agreement.id)
    UserAgreement.objects.create(agreement=agreement, user=user)

    if commit:
        user.save(update_fields=['agreements'])