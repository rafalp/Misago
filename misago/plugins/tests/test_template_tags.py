from django.template import Context, Template
from django.utils.safestring import mark_safe

from ..outlets import PluginOutlet, append_outlet_action, prepend_outlet_action


def strong_action(request, context):
    body = context.get("value") or "none"
    return mark_safe(f"<strong>{body}</strong>")


def em_action(request, context):
    body = context.get("value") or "none"
    return mark_safe(f"<em>{body}</em>")


def test_empty_outlet_renders_nothing(patch_outlets, render_outlet_template, snapshot):
    html = render_outlet_template()
    assert snapshot == html


def test_outlet_renders_appended_plugin(
    patch_outlets, render_outlet_template, snapshot
):
    append_outlet_action(PluginOutlet.TEST, strong_action)
    html = render_outlet_template()

    assert snapshot == html


def test_outlet_renders_prepended_plugin(
    patch_outlets, render_outlet_template, snapshot
):
    prepend_outlet_action(PluginOutlet.TEST, strong_action)
    html = render_outlet_template()

    assert snapshot == html


def test_outlet_renders_multiple_plugins(
    patch_outlets, render_outlet_template, snapshot
):
    append_outlet_action(PluginOutlet.TEST, strong_action)
    prepend_outlet_action(PluginOutlet.TEST, em_action)
    prepend_outlet_action(PluginOutlet.TEST, strong_action)
    html = render_outlet_template()

    assert snapshot == html


def test_outlet_renders_plugins_with_context(
    patch_outlets, render_outlet_template, snapshot
):
    append_outlet_action(PluginOutlet.TEST, strong_action)
    prepend_outlet_action(PluginOutlet.TEST, em_action)
    prepend_outlet_action(PluginOutlet.TEST, strong_action)
    html = render_outlet_template({"value": "context"})

    assert snapshot == html


def render_hasplugins_template(context: dict | None = None):
    template = Template(
        """
        {% load misago_plugins %}
        <div>{% hasplugins TEST %}plugins{% endhasplugins %}</div>
        """
    )

    return template.render(Context(context or {})).strip()


def test_hasplugins_tag_renders_nothing_if_no_plugins_exist(patch_outlets, snapshot):
    html = render_hasplugins_template()

    assert snapshot == html


def test_hasplugins_tag_renders_value_if_plugins_exist(patch_outlets, snapshot):
    append_outlet_action(PluginOutlet.TEST, strong_action)
    html = render_hasplugins_template()

    assert snapshot == html


def render_haspluginselse_template(context: dict | None = None):
    template = Template(
        """
        {% load misago_plugins %}
        <div>{% hasplugins TEST %}plugins{% else %}noplugins{% endhasplugins %}</div>
        """
    )

    return template.render(Context(context or {})).strip()


def test_hasplugins_else_tag_renders_else_if_no_plugins_exist(patch_outlets, snapshot):
    html = render_haspluginselse_template()

    assert snapshot == html


def test_hasplugins_else_tag_renders_value_if_plugins_exist(patch_outlets, snapshot):
    append_outlet_action(PluginOutlet.TEST, strong_action)
    html = render_haspluginselse_template()

    assert snapshot == html
