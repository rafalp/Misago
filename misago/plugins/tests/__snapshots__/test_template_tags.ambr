# serializer version: 1
# name: test_empty_outlet_renders_nothing
  '<div></div>'
# ---
# name: test_hasplugins_else_tag_renders_else_if_no_plugins_exist
  '<div>noplugins</div>'
# ---
# name: test_hasplugins_else_tag_renders_value_if_plugins_exist
  '<div>plugins</div>'
# ---
# name: test_hasplugins_tag_renders_nothing_if_no_plugins_exist
  '<div></div>'
# ---
# name: test_hasplugins_tag_renders_value_if_plugins_exist
  '<div>plugins</div>'
# ---
# name: test_outlet_renders_appended_plugin
  '<div><strong>none</strong></div>'
# ---
# name: test_outlet_renders_multiple_plugins
  '<div><strong>none</strong><em>none</em><strong>none</strong></div>'
# ---
# name: test_outlet_renders_plugins_with_context
  '<div><strong>context</strong><em>context</em><strong>context</strong></div>'
# ---
# name: test_outlet_renders_prepended_plugin
  '<div><strong>none</strong></div>'
# ---
